### PR TITLE
New REST API parentlocks; new WMStats CP thread for resolving parentage

### LINKED ISSUE #7651: Locking data from request manager in the view of dynamo
Opening this GH issue for starting the thread that could lead to a more stable system.
Dynamo (@yiiyama) needs to know the data that should not be touch, to the eye of Dataops. unified is taking care of this for the moment and in the way it is implemented right now, I think this would be best as a view in reqmgr2.
All inputs of workflows in status assignment-approved, acquired, assigned, running-open, running-closed, force-complete, completed, should appear in the list.
All outputs of workfows in  acquired, assigned, running-open, running-closed, force-complete, completed  should appear in the list.
All secondary of workflows in status assignment-approved, acquired, assigned, running-open, running-closed, force-complete, completed less than 2 months ago should appear in the list.

This should cover most use cases.

### PR DESCRIPTION
**Pull request by goughes**
Fixes #7651 
Replaces #9540 

#### Status
In development

#### Description
Since Alan is enjoying vacation, I picked the commits from his PR #9540 and added some additional required pieces. 
* an updated filter to only query active datasets requesting parent datasets
* a check for an empty data cache which returns a 503 error

This PR provides the following:
* a new wmstatsserver REST API `parentlocks`, which should return a list of active parent datasets in the system
* a new WMStats CherryPy thread `BuildParentLock`. Responsable for parsing the active data cached in WMStats, and running the DBS parentage discovery
* an extra `DataCache` property to store a list of parent datasets, to be served through the REST `parentlocks`

What deserves extra work are those corner cases where the data cache might be empty.
In other words, for WMStats unavailability and/or redeployment, there is a small time window that we would not have data populated in the cache, thus yielding an empty result for either the `parentlocks` or the `globallocks` REST API.

In addition to that, I also _think_ we should extend the `globallocks` to consider workflows sitting in `assignment-approved` status. (done)

#### Is it backward compatible (if not, which system it affects?)
No, it's new!

#### Related PRs
none

#### External dependencies / deployment changes
It requires these deployment changes:
https://github.com/dmwm/deployment/pull/850

### CODE REVIEW

**Comment by goughes** [comment]
I tested the empty data cache scenarios a couple ways:

1.) I removed the data cache task from the reqmon config so the cache never gets created. Queries to the API return the expected 503 error and an exception log is made in the reqmon log file.

```
[egough@lxplus754 ~]$ curl --key /tmp/x509up_u33175 --cert /tmp/x509up_u33175 -k -D - https://esg-dmwm-dev1.cern.ch/wmstatsserver/data/parentlocks
HTTP/1.1 503 Service Unavailable
Date: Fri, 21 Feb 2020 15:57:04 GMT
Server: CherryPy/17.4.0
Set-Cookie: cms-auth=09d5ba3eafd7746cbda28d3eded540aa01123e051b0bbe53f65c68e4815167879ee299ae862af389;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
X-Error-Http: 503
X-Rest-Time: 2903.938 us
X-Error-Detail: Database unavailable
X-Error-Id: 04c618c2e2979318af78c4683ef96000
X-Rest-Status: 200
Content-Type: text/html;charset=utf-8
CMS-Server-Time: D=24190 t=1582300624676517
Connection: close
Transfer-Encoding: chunked
```
2.) I re-enabled the data cache thread and only started the frontend and reqmon. The data cache thread errored because it couldn't talk to couchdb. I again got the expected error message.

```
[egough@lxplus754 ~]$ curl --key /tmp/x509up_u33175 --cert /tmp/x509up_u33175 -k -D - https://esg-dmwm-dev1.cern.ch/wmstatsserver/data/parentlocks
HTTP/1.1 503 Service Unavailable
Date: Fri, 21 Feb 2020 16:16:34 GMT
Server: CherryPy/17.4.0
Set-Cookie: cms-auth=fd8acf74393bf18f1a3972361c93a7f88cb4d11d16c46116d093749dba73112e29fb00543179d0a6;path=/;secure;httponly;expires=Thu, 01-Jan-1970 00:00:01 GMT
X-Error-Http: 503
<snip>
```
Then I started the rest of the services and waited a bit for the data cache and parent threads to run.
```
INFO:dataCacheTasks:DataCache is updated: 13
INFO:parentLock:Executing parent lock cherrypy thread
INFO:parentLock:Found 0 unique datasets requiring the parent dataset
INFO:parentLock:Parentage lookup complete and cache renewed
```
Queries to the API worked as expected.
```
[egough@lxplus754 ~]$ curl --key /tmp/x509up_u33175 --cert /tmp/x509up_u33175 -k -D - https://esg-dmwm-dev1.cern.ch/wmstatsserver/data/parentlocks
HTTP/1.1 200 OK
```

**Comment by goughes** [comment]
I re-used a DatabaseUnavailable exception for this. I could make a data cache specific rest exception if necessary.

**Comment by goughes** [comment]
Those two pylint errors are from code in the REST Errors.py I did not change.

**Review by todor-ivanov — APPROVED** [review]
Following our discussion with Erik from earlier today, which helped me understand all the details and given that all requested changes were provided I can say the PR is well done.

File: src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py [hunk 1]
```diff
@@ -144,12 +144,16 @@
...
                                           'running-open', 'running-closed', 'force-complete',
                                           'completed', 'closed-out']}
-ACTIVE_NO_CLOSEOUT_FILTER = {"RequestStatus": ['assigned', 'staging', 'staged', 'acquired',
+ACTIVE_NO_CLOSEOUT_FILTER = {"RequestStatus": ['assignment-approved', 'assigned', 'staging', 'staged', 'acquired',
                                                'running-open', 'running-closed', 'force-complete',
                                                'completed']}
+# used to determine active workflows requiring parent datasets
+ACTIVE_NO_CLOSEOUT_PARENT_FILTER = {"RequestStatus": ['assignment-approved', 'assigned', 'staging', 'staged', 'acquired',
+                                               'running-open', 'running-closed', 'force-complete',
+                                               'completed'], "IncludeParents": True}
```
Comment (amaltaro) [inline src/python/WMCore/ReqMgr/DataStructs/RequestStatus.py]: I think this `IncludeParents` filter doesn't belong to this module. We can get back to it in the future though

**Review by amaltaro — COMMENTED** [review]
Thanks for following this up, Erik. Changes look good to me. I had only one minor comment, no need to change anything for now though.

While reviewing these changes, I think we might still have one case that would give empty results, even though it's just lack of data. E.g.: cache is not empty, but the BuildParentLocks thread still needs to run and build a list of parent datasets. When requesting `parentlocks` API, it would return an empty list, right? Even though there might be parent datasets to be locked. I don't know how to deal with that use case though... and we can likely get back to it later in march as well.
